File: Prácticas/P3/Java/civitas/src/JuegoTexto/VistaTextual.java
package JuegoTexto;

import civitas.CivitasJuego;
import civitas.Diario;
import civitas.OperacionesJuego;
import civitas.SalidasCarcel;
import civitas.Respuestas;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;
import civitas.Casilla;
import civitas.Jugador;
import civitas.OperacionesInmobiliarias;
import civitas.TituloPropiedad;

class VistaTextual {
  
  CivitasJuego juegoModel;
  private int iGestion=-1;
  private int iPropiedad=-1;
  private static String separador = "=====================";
  
  private Scanner in;
  
  VistaTextual () {
    in = new Scanner (System.in);
  }
  
  void mostrarEstado(String estado) {
    System.out.println (estado);
  }
              
  void pausa() {
    System.out.print ("Pulsa una tecla");
    in.nextLine();
  }

  int leeEntero (int max, String msg1, String msg2) {
    Boolean ok;
    String cadena;
    int numero = -1;
    do {
      System.out.print (msg1);
      cadena = in.nextLine();
      try {  
        numero = Integer.parseInt(cadena);
        ok = true;
      } catch (NumberFormatException e) { // No se ha introducido un entero
        System.out.println (msg2);
        ok = false;  
      }
      if (ok && (numero < 0 || numero >= max)) {
        System.out.println (msg2);
        ok = false;
      }
    } while (!ok);

    return numero;
  }

  int menu (String titulo, ArrayList<String> lista) {
    String tab = "  ";
    int opcion;
    System.out.println (titulo);
    for (int i = 0; i < lista.size(); i++) {
      System.out.println (tab+i+"-"+lista.get(i));
    }

    opcion = leeEntero(lista.size(),
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo");
    return opcion;
  }

  SalidasCarcel salirCarcel() {
    int opcion = menu ("Elige la forma para intentar salir de la carcel",
      new ArrayList<> (Arrays.asList("Pagando","Tirando el dado")));
    return (SalidasCarcel.values()[opcion]);
  }

  Respuestas comprar() {
    int opcion = menu (juegoModel.getCasillaActual().toString()+ "\n¿Desea comprar la calle?",
        new ArrayList<> (Arrays.asList("Si", "No")));
    return (Respuestas.values()[opcion]);
  }

  void gestionar() {
    iGestion = menu ("Eliga la gestion que desea hacer",
        new ArrayList<> (Arrays.asList("Vender", "Hipotecar", "Cancelar Hipoteca", "Construir Casa", "Construir Hotel", "Terminar")));
    
    ArrayList<String> calles = new ArrayList<String>();
    
    for(int i=0; i<juegoModel.getJugadorActual().getPropiedades().size(); i++){
        calles.add(juegoModel.getJugadorActual().getPropiedades().get(i).getNombre());
    }
    iPropiedad = menu("¿Sobre que casilla desea operar?",calles);
  }
  
  public int getGestion(){
      return iGestion;
  }
  
  public int getPropiedad(){
      return iPropiedad;
  }   

  void mostrarSiguienteOperacion(OperacionesJuego operacion) {
      switch(operacion){
          case AVANZAR:
              System.out.println("La siguiente operacion es avanzar");
              break;
          case COMPRAR:
              System.out.println("La siguiente operacion es comprar");
              break;
          case GESTIONAR:
              System.out.println("La siguiente operacion es gestionar");
              break;
          case PASAR_TURNO:
              System.out.println("La siguiente operacion es pasar turno");
              break;
          default:
              System.out.println("La siguiente operacion es salir de la carcel");        
      }
  }

  void mostrarEventos() {
      while(Diario.getInstance().eventosPendientes()){
          System.out.println(Diario.getInstance().leerEvento());
      }
  }
  
  public void setCivitasJuego(CivitasJuego civitas){ 
        juegoModel=civitas;
    }
  
  public void actualizarVista(){
      System.out.print(juegoModel.infoJugadorTexto());
  } 
}

File: Prácticas/P3/Ruby/civitas/lib/vista_textual.rb
#encoding:utf-8
require_relative 'operaciones_juego.rb'
require_relative 'civitas_juego.rb'
require_relative 'salidas_carcel.rb'
require_relative 'respuestas.rb'
require_relative 'diario.rb'
require_relative 'casilla.rb'
require_relative 'operaciones_inmobiliarias.rb'
require_relative 'jugador.rb'
require_relative 'titulo_propiedad.rb'
require 'io/console'

module JuegoTexto

  class Vista_textual
    include Civitas
    attr_reader :i_gestion, :i_propiedad, :juego_model
    
    def initialize
      @i_gestion = -1
      @i_propiedad = -1
    end
    
    def mostrar_estado(estado)
      puts estado
    end

    
    def pausa
      print "Pulsa una tecla"
      STDIN.getch
      print "\n"
    end

    def lee_entero(max,msg1,msg2)
      ok = false
      begin
        print msg1
        cadena = gets.chomp
        begin
          if (cadena =~ /\A\d+\Z/)
            numero = cadena.to_i
            ok = true
          else
            raise IOError
          end
        rescue IOError
          puts msg2
        end
        if (ok)
          if (numero >= max)
            ok = false
          end
        end
      end while (!ok)

      return numero
    end



    def menu(titulo,lista)
      tab = "  "
      puts titulo
      index = 0
      lista.each { |l|
        puts tab+index.to_s+"-"+l
        index += 1
      }

      opcion = lee_entero(lista.length,
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo")
      return opcion
    end
    
    def salir_carcel
      lista_respuestas = [SalidasCarcel::PAGANDO,SalidasCarcel::TIRANDO]
      opcion = menu("Elige la forma para intentar salir de la carcel",["PAGANDO","TIRANDO"])
      lista_respuestas[opcion]
    end
    
    def comprar
      lista_respuestas = [Respuestas::NO,Respuestas::SI]
      opcion = menu(@juego_model.get_casilla_actual.to_s + "\n¿Desea comprar la calle?",["NO","SI"])
      lista_respuestas[opcion]
    end

    def gestionar
      lista_respuestas = [OperacionesInmobiliarias::VENDER,OperacionesInmobiliarias::HIPOTECAR,OperacionesInmobiliarias::CANCERLARHIPOTECA,OperacionesInmobiliarias::CONSTRUIRCASA,OperacionesInmobiliarias,OperacionesInmobiliarias::CONSTRUIRHOTEL,OperacionesInmobiliarias::TERMINAR]
      opcion = menu("Eliga la gestion que desea hacer",["Vender", "Hipotecar", "Cancelar Hipoteca", "Construir Casa", "Construir Hotel", "Terminar"])
      @i_gestion = lista_respuestas[opcion]
      
      nombres = Array.new
      
      for i in 0..@juego_model.get_jugador_actual.propiedades.length
        nombres << @juego_model.get_jugador_actual.propiedades.at(i).nombre
      end
      @i_propiedad = menu("¿Sobre que casilla deseas operar?",nombres)
    end

    def mostrarSiguienteOperacion(operacion)
      case operacion
      when OperacionesJuego::AVANZAR
        puts "La siguiente operacion es avanzar"
      when OperacionesJuego::COMPRAR
        puts "La siguiente operacion es comprar"
      when OperacionesJuego::GESTIONAR
        puts "La siguiente operacion es gestionar"
      when OperacionesJuego::PASARTURNO
        puts "La siguiente operacion es pasar turno"
      else
        puts "La siguiente operacion es salir de la carcel"
      end
    end

    def mostrarEventos
      while Diario.instance.eventos_pendientes
        puts Diario.instance.leer_evento
      end
    end

    def setCivitasJuego(civitas)
         @juego_model = civitas
    end

    def actualizarVista
      puts @juego_model.info_jugador_texto
    end

    
  end

end
